tests/end_to_end/netsh_test.cpp

```diff
@@ -273,29 +273,117 @@ TEST_CASE("show verification droppacket_unsafe.o", "[netsh][verification]")
-    REQUIRE(object != nullptr);
-    REQUIRE(program_fd != -1);
-
-    std::string output = _run_netsh_command(handle_ebpf_show_programs, L"xdp", nullptr, nullptr, &result);
+    int result;
+    std::string output =
+        _run_netsh_command(handle_ebpf_add_program, L"tail_call.o", L"pinned=mypinname", nullptr, &result);
+    REQUIRE(strcmp(output.c_str(), "Loaded with ID 196609\n") == 0);
+    REQUIRE(result == NO_ERROR);
+
+    // Show programs in normal (table) format.
+    output = _run_netsh_command(handle_ebpf_show_programs, L"xdp", nullptr, nullptr, &result);
+    REQUIRE(result == NO_ERROR);
+    REQUIRE(
+        output == "\n"
+                  "    ID  Pins  Links  Mode       Name\n"
+                  "======  ====  =====  =========  ====================\n"
+                  "196609     1      1  JIT        caller\n"
+                  "262145     0      0  JIT        callee\n");
+
+    // Test filtering by "attached=yes".
+    output = _run_netsh_command(handle_ebpf_show_programs, L"attached=yes", nullptr, nullptr, &result);
+    REQUIRE(result == NO_ERROR);
+    REQUIRE(
+        output == "\n"
+                  "    ID  Pins  Links  Mode       Name\n"
+                  "======  ====  =====  =========  ====================\n"
+                  "196609     1      1  JIT        caller\n");
+
+    // Test filtering by "attached=no".
+    output = _run_netsh_command(handle_ebpf_show_programs, L"attached=no", nullptr, nullptr, &result);
     REQUIRE(result == NO_ERROR);
+    REQUIRE(
+        output == "\n"
+                  "    ID  Pins  Links  Mode       Name\n"
+                  "======  ====  =====  =========  ====================\n"
+                  "262145     0      0  JIT        callee\n");
 
+    // Test filtering by "pinned=yes".
+    output = _run_netsh_command(handle_ebpf_show_programs, L"pinned=yes", nullptr, nullptr, &result);
+    REQUIRE(result == NO_ERROR);
     REQUIRE(
         output == "\n"
-                  "    ID            File Name         Section             Name      Mode\n"
-                  "====== ==================== =============== ================ =========\n"
-                  "196609          tail_call.o        xdp_prog           caller       JIT\n"
-                  "262145          tail_call.o      xdp_prog/0           callee       JIT\n");
+                  "    ID  Pins  Links  Mode       Name\n"
+                  "======  ====  =====  =========  ====================\n"
+                  "196609     1      1  JIT        caller\n");
 
-    bpf_object__close(object);
+    // Test filtering by "pinned=no".
+    output = _run_netsh_command(handle_ebpf_show_programs, L"pinned=no", nullptr, nullptr, &result);
+    REQUIRE(result == NO_ERROR);
+    REQUIRE(
+        output == "\n"
+                  "    ID  Pins  Links  Mode       Name\n"
+                  "======  ====  =====  =========  ====================\n"
+                  "262145     0      0  JIT        callee\n");
```

Comment:
Should the unpinned program not get unloaded? or it is because netsh has not exited in the test case, so the program fd is still not closed?

Reply:
Correct, netsh still holds the object reference.   Each _run_netsh_command() call is logically a separate command inside the same netsh process in interactive mode